### PR TITLE
[rocjitsu] Preserve LDS contents on reuse and honor queue CU masks

### PR DESCRIPTION
**Pull request by kuhar**
A downstream LDS reuse test read zeros because workgroup allocations cleared LDS, while SET_CU_MASK returned success without constraining placement. Retain CU and paired-WGP backing contents across reuse and honor queue masks in local and daemon execution.

Map KFD mask bits to physical CUs, including RDNA sibling pairs, and restrict ordinary and clustered placement to the selected units. Split multi-XCD grids only among eligible XCDs while preserving completion on the queue owner. Keep queue and dispatch mask snapshots in owned vectors and defer queues with an empty mask until they are enabled again.

Cover LDS retention, physical CU selection, masked streams, clustered dispatches, and malformed ioctl payloads. The original local and daemon reproducers complete, and all 56 focused tests pass. Physical gfx1201 probes retain all 64 KiB across 100 pinned dispatch pairs.

### CODE REVIEW

**Review by atgutier — COMMENTED** [review]
Some agent comments:

[MAJOR] /tmp/pr11491-review.Wz1APH/repo/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/
  command_processor.cpp:2458 — An empty CU mask also blocks CP-only packets

  The unconditional early return happens before decoding the AQL packet. Consequently, barrier,
  barrier-value, and PM4 IB packets at the queue head are never consumed, even though they execute no
  shader and do not require an enabled CU. Their completion signals can remain stuck until the queue
  receives a non-empty mask.

  KFD implements the mask through the MQD static-thread-management fields; it does not suspend the
  command processor or deactivate the queue.

  Suggested fix: Decode the head packet before applying empty-mask deferral. Defer kernel-dispatch
  packets without consuming them, but continue processing non-kernel packets. Add a test with an empty
  mask and a head barrier carrying a completion signal.

  Confidence: High.

  [MAJOR] /tmp/pr11491-review.Wz1APH/repo/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/
  simulated_kfd.cpp:2446 — Invalid RDNA CU masks are silently widened

  For RDNA, the implementation considers a pair enabled when either bit is set and then enables both
  physical CUs. Real KFD explicitly rejects 01 and 10 pairs with -EINVAL; see /home/agutierr/rocm-dev/
  amdgpu/drivers/gpu/drm/amd/amdkfd/kfd_process_queue_manager.c:647.

  The new placement cases at /tmp/pr11491-review.Wz1APH/repo/emulation/rocjitsu/tests/
  kfd_ioctl_test.cpp:403 use bits 9 and 19 individually and therefore validate behavior the kernel
  rejects. Direct KFD clients would see an invalid request succeed and execute on an additional CU
  they did not enable.

  Suggested fix: Before changing queue state, validate every RDNA pair as either 00 or 11. Update
  placement tests to use valid paired masks and add explicit rejection tests for both partial-pair
  forms.

  Confidence: High.


  [MAJOR] /tmp/pr11491-review.Wz1APH/repo/emulation/rocjitsu/tests/CMakeLists.txt:2976 — The new HIP
  test makes sanitizer CI fail

  HipMemcpyTest.MaskedStreamLdsReuse and its daemon wrapper are registered under ASan and TSan.
  Current CI shows:

  - ASan/UBSan: local and daemon cases fail with 5,712-byte LeakSanitizer reports originating in
    libhsa-runtime64.

  - TSan: the local case fails with six reports involving the uninstrumented runtime’s host writes and
    GpuMemory::read_mapped.

  The nearby CMake logic already documents and excludes another HIP memcpy case for these exact
  external ROCr sanitizer limitations.

  Suggested fix: Gate both the local registration and daemon registration (/tmp/pr11491-review.Wz1APH/
  repo/emulation/rocjitsu/tests/CMakeLists.txt:3108) with NOT RJ_ENABLE_ASAN AND NOT RJ_ENABLE_TSAN,
  or redesign the test so it does not traverse the problematic runtime path.

  Confidence: High.

  [MINOR] /tmp/pr11491-review.Wz1APH/repo/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/
  simulated_kfd.cpp:2416 — Missing queue returns the wrong KFD errno

  The simulator returns -EINVAL when queue_id is unknown. Kernel pqm_update_mqd() returns -EFAULT for
  that condition at /home/agutierr/rocm-dev/amdgpu/drivers/gpu/drm/amd/amdkfd/
  kfd_process_queue_manager.c:637. The new local and daemon tests currently codify the incompatible
  -EINVAL result.

  Suggested fix: Return -EFAULT and update both tests.

  Confidence: High.

**Comment by kuhar** [comment]
@atgutier addressed your comments

**Review by atgutier — COMMENTED** [review]
Agent found another major issue:

[MAJOR] /tmp/pr11491-review.Wz1APH/repo/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/
  command_processor.cpp:768 — Changing a queue’s CU mask can violate cross-XCD barrier ordering

  Reasoning: Selective fanout sends a kernel entry only to XCDs containing currently enabled CUs.
  Consider a long-running dispatch A masked to XCD 0, followed by changing the queue mask to XCD 7 and
  submitting barrier-bit dispatch B. XCD 7 never received an entry for A, so B is first in its replica
  queue. barrier_satisfied() (/tmp/pr11491-review.Wz1APH/repo/emulation/rocjitsu/lib/rocjitsu/src/
  rocjitsu/vm/amdgpu/command_processor.cpp:1332) treats index zero as having no predecessors and
  allows B to execute while A is still running. The owner’s correctly ordered entry list cannot
  prevent the already-delivered peer shard from starting.

  The same missing-history problem can bypass an explicit replicated barrier packet: an XCD omitted
  from A sees the barrier without A in front of it and can retire the barrier early.

  Evidence:

  - fan_out_dispatch() (/tmp/pr11491-review.Wz1APH/repo/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/
    vm/amdgpu/command_processor.cpp:784) sends shards only to participants.

  - barrier_satisfied() (/tmp/pr11491-review.Wz1APH/repo/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/
    vm/amdgpu/command_processor.cpp:1332) considers only entries in the local replica.

  - Shards are delivered during packet fetch, before owner-side ordered processing, at /tmp/pr11491-
    review.Wz1APH/repo/emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/
    command_processor.cpp:2389.

  - The class’s existing invariant explicitly requires every XCD to receive an entry, including empty
    shares, to preserve ordering at /tmp/pr11491-review.Wz1APH/repo/emulation/rocjitsu/lib/rocjitsu/
    src/rocjitsu/vm/amdgpu/command_processor.h:452.

  - The added fanout test uses the same mask for both dispatches at /tmp/pr11491-review.Wz1APH/repo/
    emulation/rocjitsu/tests/xcd_distribution_test.cpp:281, so it does not cover a mask transition.

  Suggested fix: Send every XCD a dispatch entry. Give excluded XCDs zero-work placeholder shards
  sharing the same GridCompletion, while distributing real work only among eligible participants.
  Alternatively, introduce a shared predecessor token independent of local entry history. Add a
  regression with dispatch A masked to one XCD, then change to a disjoint XCD and submit barrier-bit
  dispatch B; verify B cannot begin before A completes in both single-threaded and thread-per-XCD
  modes.

  Confidence: High.

**Review by atgutier — APPROVED** [review]
I think this looks good once you fix the agent's major finding. So approving on the condition you fix that.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/simulated_kfd.cpp [hunk 1]
```diff
@@ -2396,6 +2398,78 @@ int SimulatedKfd::create_queue_ioctl(KfdProcess &proc, void *arg) {
...
+    return -EFAULT;
+
+  uint32_t gpu_id = 0;
+  {
+    std::lock_guard<std::mutex> lock(proc.alloc_mutex_);
+    const auto queue = proc.queue_snapshot_map_.find(args->queue_id);
+    if (queue == proc.queue_snapshot_map_.end())
+      return -EFAULT;
+    const uint32_t type = queue->second.queue_type;
+    if (type != KFD_IOC_QUEUE_TYPE_COMPUTE && type != KFD_IOC_QUEUE_TYPE_COMPUTE_AQL)
```
Comment (atgutier) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/simulated_kfd.cpp]: I think the SDMA can support cu masks.

**Review by atgutier — APPROVED** [review]
Approving with a minor comment.
